constraint/pkg/client/client.go

```diff
@@ -1,355 +1,183 @@
 package client
 
 import (
-	"bytes"
 	"context"
-	"errors"
 	"fmt"
-	"path"
 	"sort"
 	"strings"
 	"sync"
 
-	"github.com/open-policy-agent/frameworks/constraint/pkg/apis/constraints"
+	apiconstraints "github.com/open-policy-agent/frameworks/constraint/pkg/apis/constraints"
 	"github.com/open-policy-agent/frameworks/constraint/pkg/client/crds"
 	"github.com/open-policy-agent/frameworks/constraint/pkg/client/drivers"
-	"github.com/open-policy-agent/frameworks/constraint/pkg/client/drivers/local"
 	clienterrors "github.com/open-policy-agent/frameworks/constraint/pkg/client/errors"
-	"github.com/open-policy-agent/frameworks/constraint/pkg/client/regolib"
-	constraintlib "github.com/open-policy-agent/frameworks/constraint/pkg/core/constraints"
 	"github.com/open-policy-agent/frameworks/constraint/pkg/core/templates"
 	"github.com/open-policy-agent/frameworks/constraint/pkg/handler"
 	"github.com/open-policy-agent/frameworks/constraint/pkg/types"
-	"github.com/open-policy-agent/opa/format"
 	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
-	"k8s.io/apimachinery/pkg/runtime/schema"
 )
 
-type templateEntry struct {
-	template *templates.ConstraintTemplate
-	CRD      *apiextensions.CustomResourceDefinition
-	Targets  []string
-}
-
+// Client tracks ConstraintTemplates and Constraints for a set of Targets.
+// Allows validating reviews against Constraints.
+//
+// Threadsafe.
+// Assumes concurrent calls for the same object do not happen. For example -
```

Comment:
This note is completely unrelated to parallel compilation of Templates. Removing parallel computation of Templates does not remove this requirement.

This is a requirement as it is impossible for Frameworks to handle safely if this assumption is violated. There is _absolutely nothing_* that Frameworks can do to ensure proper behavior in the case that a caller sends a conflicting request to change an object before the previous operation has succeeded. Frameworks can't detect this scenario:

1. Start thread for `RemoveTemplate`
2. Thread for `RemoveTemplate` is frozen before any locking occurs (this can be the case even if the mutex lock is on the first line - e.g. while the function's stack is being instantiated)
3. Start thread for `AddTemplate`
4. `Add Template` gets a read lock.
5. Thread for `RemoveTemplate` wakes up, but can't do anything because it cannot acquire a lock.
6. `AddTemplate` completes successfully and releases the lock.
7. `RemoveTemplate` runs and deletes the Template.

The only safe way to handle this problem is to tell callers that they must make this guarantee.

*Under the current design. An alternative design would have Client run in its own thread, accepting work items synchronously, acquiring resource-specific locks, and launching work in goroutines, blocking callers until the work item is processed. This would be a vast departure from the design we currently have.

Reply:
That makes sense. My main concern would be that concurrent requests cannot put the framework into an inconsistent state (e.g., parts of its state think the template has been added, parts of its state think it's been removed).

Clients should be responsible for avoiding race conditions in their own logic (which is what you describe), we should be responsible for maintaining consistent state (or raising a panic when we've failed to do so), regardless of how we're called.

Reply:
Changed Client to not allow concurrent updates or compilation. I don't want to hold up the PR on this - speeding up compilation by 20x is significant on it's own. While it would be nice to get the 5x additional speedup to 100x, we can do that later (and do it right).